include/parsing/warc.hpp

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <cctype>
+#include <iostream>
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+#include <gsl/span>
+
+namespace ds2i {
+
+using Field_Map = std::unordered_map<std::string, std::string>;
+
+class Warc_Format_Error : public std::exception {
+    std::string message_;
+
+   public:
+    Warc_Format_Error(std::string line, std::string message) : message_(message + line) {}
+    [[nodiscard]] char const *what() const noexcept override { return message_.c_str(); }
+};
+
+class Warc_Record;
+std::istream &read_warc_record(std::istream &in, Warc_Record &record);
+
+//! A WARC record.
+class Warc_Record {
+   private:
+    std::string version_;
+    Field_Map   warc_fields_;
+    Field_Map   http_fields_;
+    std::string content_;
+
+   public:
+    Warc_Record() = default;
+    explicit Warc_Record(std::string version) : version_(std::move(version)) {}
+    [[nodiscard]] auto type() const -> std::string const & { return warc_fields_.at("warc-type"); }
+    [[nodiscard]] auto valid() const noexcept -> bool { return type() == "response"; }
+    [[nodiscard]] auto warc_content_length() const -> std::size_t {
+        auto &field_value = warc_fields_.at("content-length");
+        try {
+            return std::stoi(field_value);
+        } catch (std::invalid_argument &error) {
+            throw Warc_Format_Error(field_value, "could not parse content length: ");
+        }
+    }
+    [[nodiscard]] auto http_content_length() const -> std::size_t {
+        auto &field_value = http_fields_.at("content-length");
+        try {
+            return std::stoi(field_value);
+        } catch (std::invalid_argument &error) {
+            throw Warc_Format_Error(field_value, "could not parse content length: ");
+        }
+    }
+    [[nodiscard]] auto content() -> std::string & { return content_; }
+    [[nodiscard]] auto content() const -> std::string const & { return content_; }
+    [[nodiscard]] auto url() const -> std::string const & {
+        return warc_fields_.at("warc-target-uri");
+    }
+    [[nodiscard]] auto trecid() const -> std::string const & {
+        return warc_fields_.at("warc-trec-id");
```

Comment:
I realized this is non standard. When I tied to parse a CC-NEWS collection it was not able to parse any documents.

```
WARC/1.0
WARC-Record-ID: <urn:uuid:4e7b712c-f7fd-4e31-81ad-b6b8b9a190c8>
Content-Length: 38840
WARC-Date: 2016-12-01T13:28:43Z
WARC-Type: response
WARC-Target-URI: http://www.banker.bg/finansov-dnevnik/read/parlamentut-prie-biudjetite-na-vsichki-ministerstva?utm_source=rss&utm_medium=click&utm_campaign=rss
Content-Type: application/http; msgtype=response
WARC-Payload-Digest: sha1:NKWIH3RIPHRCTJ7F7VM45KZFUR5DCXMW
WARC-Block-Digest: sha1:EOKDPEU3LQ7XESEPNRQO34YDAEDADIL6
```

Reply:
Just open a new issue and copy it there.

It makes sense that TREC ID is not a WARC standard. But that begs some questions:

- Is there a "CC-NEWS" ID by which you can identify a specific document?
- For that matter, are there any evaluation queries we could use, say, to test precision?

Reply:
#55